### PR TITLE
Implement OCI image puller binary and writer utility

### PR DESCRIPTION
**Pull request by xwinxu**
Extended rules_docker API to now contain the `Write` function which outputs the images from the puller in OCI standard format. BUILD files updated in the process.

### CODE REVIEW

File: container/go/cmd/puller/puller.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2015 The Bazel Authors. All rights reserved.
+/// Copyright 2015 The Bazel Authors. All rights reserved.
```
Comment (xingao267) [inline container/go/cmd/puller/puller.go]: Double check for these kind of unnecessary changes.

File: container/go/cmd/puller/puller.go [hunk 2]
```diff
@@ -41,67 +43,64 @@ var (
...
-			log.Fatal("ref wasn't a tag or digest")
-		}
-		s := fmt.Sprintf("%s:%s", d.Repository.Name(), iWasADigestTag)
-		tag, err = name.NewTag(s)
-		if err != nil {
-			log.Fatalf("parsing digest as tag (%s): %v", s, err)
-		}
+	// // Image file to write to disk.
+	if err := oci.Write(img, dstPath); err != nil {
+		// if err := oci.Write(img, path); err != nil {
```
Comment (xingao267) [inline container/go/cmd/puller/puller.go]: remove

File: container/go/cmd/puller/puller.go [hunk 3]
```diff
@@ -41,67 +43,64 @@ var (
...
+	if *clientConfigDir != "" {
+		ospkg.Setenv("DOCKER_CONFIG", *clientConfigDir)
+	}
+
+	// Create a Platform struct with arguments
+	platform := v1.Platform{
+		Architecture: *arch,
+		OS:           *os,
+		OSVersion:    *osVersion,
+		OSFeatures:   strings.Fields(*osFeatures),
```
Comment (xingao267) [inline container/go/cmd/puller/puller.go]: strings.Fields assumes the input strings are separated by space. Specify this explicitly in the flag documentation for osFeatures and features.

File: container/go/pkg/oci/write.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+func Write(img v1.Image, dstPath string) error {
+	// Path represents an OCI image layout ooted in a file system path
+	var path layout.Path
+	var err error
+
+	// Open the layout, if it already exists.
+	if path, err = layout.FromPath(dstPath); err != nil {
+		// Does not already exist, so initialize it with an empty index.
+		if path, err = layout.Write(dstPath, empty.Index); err != nil {
+			log.Fatalf("cannot initialize layout: %v", err)
```
Comment (xingao267) [inline container/go/pkg/oci/write.go]: Return the error 

return fmt.Errorf(...)

File: container/go/pkg/oci/write.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+)
+
+// Write writes a v1.Image to the Path and updates the index.json to reference it.
+// This is just syntactic sugar wrapping Path.AppendImage from the go-container registry.
+func Write(img v1.Image, dstPath string) error {
+	// Path represents an OCI image layout ooted in a file system path
+	var path layout.Path
+	var err error
```
Comment (xingao267) [inline container/go/pkg/oci/write.go]: You don't need to do this.

You can just initialize err inline in your first call by changing = to :=

File: container/go/pkg/oci/write.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	if path, err = layout.FromPath(dstPath); err != nil {
+		// Does not already exist, so initialize it with an empty index.
+		if path, err = layout.Write(dstPath, empty.Index); err != nil {
+			log.Fatalf("cannot initialize layout: %v", err)
+		}
+
+	}
+
+	if err := path.AppendImage(img); err != nil {
+		return fmt.Errorf("unable to write image to path")
```
Comment (xingao267) [inline container/go/pkg/oci/write.go]: Include the original err

return fmt.Errorf("unable to write image to path: %v", err)

Comment (xingao267) [inline container/go/pkg/oci/write.go]: Include the original err

return fmt.Errorf("unable to write image to path: %v", err)

Also use errors.Wrapf everywhere. I think that's more recommended.

File: container/go/pkg/oci/write.go (hunk 5 above)
Comment (xwinxu) [reply container/go/pkg/oci/write.go]: I tried that before. But it gives me a path undefined error because I reuse path again in line 43.

File: container/go/cmd/puller/BUILD [hunk 7]
```diff
@@ -11,25 +11,31 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+# ###
+# Build file for new puller binary based on go-containerregistry backend.
```
Comment (xingao267) [inline container/go/cmd/puller/BUILD]: Usually we don't write this kind of comment. But it doesn't hurt to have it. No need to change.
